src/Symfony/Component/EventDispatcher/CompiledEventDispatcher.php

```diff
@@ -0,0 +1,241 @@
+ *     Especially when dispatching the KernelEvents::REQUEST event, this leads
+ *     to a more timely invocation of the first listener. Overall dispatch
+ *     runtime is not affected by this change though.
+ *   </dd>
+ * </dl>
+ */
+class CompiledEventDispatcher implements EventDispatcherInterface
+{
+    /**
+     * The service container.
+     *
+     * @var \Symfony\Component\DependencyInjection\IntrospectableContainerInterface
+     */
+    private $container;
+
+    /**
+     * Listener definitions.
+     *
+     * A nested array of listener definitions keyed by event name and priority.
+     * A listener definition is an associative array with one of the following key
+     * value pairs:
+     * - callable: A callable listener
+     * - service: An array of the form array(service id, method)
+     *
+     * A service entry will be resolved to a callable only just before its
+     * invocation.
+     *
+     * @var array
+     */
+    private $listeners;
+
+    /**
+     * Whether listeners need to be sorted prior to dispatch, keyed by event name.
+     *
+     * @var array
+     */
+    private $unsorted = array();
+
+    /**
+     * Constructs a container aware event dispatcher.
+     *
+     * @param \Symfony\Component\EventDispatcher\IntrospectableContainerInterface $container
+     *   The service container.
+     * @param array $listeners
+     *   A nested array of listener definitions keyed by event name and priority.
+     *   The array is expected to be ordered by priority. A listener definition is
+     *   an associative array with one of the following key value pairs:
+     *   - callable: A callable listener
+     *   - service: An array of the form array(service id, method)
+     *   A service entry will be resolved to a callable only just before its
+     *   invocation.
+     */
+    public function __construct(IntrospectableContainerInterface $container, array $listeners = array())
+    {
+        $this->container = $container;
+        $this->listeners = $listeners;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function dispatch($eventName, Event $event = null)
+    {
+        if ($event === NULL) {
+            $event = new Event();
+        }
+
+        $event->setDispatcher($this);
+        $event->setName($eventName);
+
+        if (isset($this->listeners[$eventName])) {
+            // Sort listeners if necessary.
+            if (isset($this->unsorted[$eventName])) {
+                krsort($this->listeners[$eventName]);
+                unset($this->unsorted[$eventName]);
+            }
+
+            // Invoke listeners and resolve callables if necessary.
+            foreach ($this->listeners[$eventName] as &$definitions) {
+                foreach ($definitions as $key => &$definition) {
+                    if (!isset($definition['callable'])) {
+                        $definition['callable'] = array($this->container->get($definition['service'][0]), $definition['service'][1]);
```

Comment:
instead of ``['service'][0]`` and ``['service'][1]``, I would prefer using 2 keys ``service`` and ``method``. Otherwise, using ``service`` as key is wrong, because it does not represent a service